### PR TITLE
Vet clinic database: add 'join table' for visits

### PR DESCRIPTION
**Pull request by SSEKPIUS**
I am required to add some many-to-many relationships and write more complex queries to use them to answer questions.
Screenshot 2023-02-03 at 9 00 50 AM
Screenshot 2023-02-03 at 9 02 45 AM
Screenshot 2023-02-03 at 9 04 53 AM
Screenshot 2023-02-03 at 9 43 47 AM
Screenshot 2023-02-03 at 9 50 13 AM
Screenshot 2023-02-03 at 9 53 45 AM
Screenshot 2023-02-03 at 10 37 48 AM
Screenshot 2023-02-03 at 10 38 18 AM
Screenshot 2023-02-03 at 10 39 04 AM
Screenshot 2023-02-03 at 10 39 49 AM
Screenshot 2023-02-03 at 10 40 26 AM
Screenshot 2023-02-03 at 10 42 51 AM
Screenshot 2023-02-03 at 11 01 46 AM

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -4,4 +4,23 @@ USE vet_clinic;
...
+CREATE TABLE species ( id BIGSERIAL NOT NULL PRIMARY KEY, name VARCHAR(255));
+-- Modify animals table:
+ALTER TABLE animals ADD COLUMN species_id INTEGER, ADD COLUMN owner_id INTEGER, DROP COLUMN species;
+ALTER TABLE animals ADD CONSTRAINT fk_species_id FOREIGN KEY (species_id) REFERENCES species (id);
+ALTER TABLE animals ADD CONSTRAINT fk_owner_id FOREIGN KEY (owner_id) REFERENCES owners (id);
+ALTER TABLE animals ADD COLUMN species varchar(255);
+
+/* Vet clinic database: add "join table" for visits */
+-- Create a table named vets with the following column
+CREATE TABLE vets (id BIGSERIAL NOT NULL PRIMARY KEY, name VARCHAR(255), age INT, date_of_graduation DATE);
```
Comment (t-yanick) [inline schema.sql]: **[ OPTIONAL]** 
- You can consider changing the data type of `id` to `int`. `Bigserial` is mostly used when you will have over 2 billion identifires in the life time of the database. This works just so well ` id int NOT NULL IDENTITY(1,1),`

File: schema.sql [hunk 2]
```diff
@@ -4,4 +4,23 @@ USE vet_clinic;
...
+-- Create a "join table" called specializations to handle this relationship btn species and vets
+CREATE TABLE specializations (species_id INTEGER REFERENCES species(id), vet_id INTEGER REFERENCES vets(id), PRIMARY KEY (species_id, vet_id));
```
Comment (t-yanick) [inline schema.sql]: - [ ] `species_id` and `vet_id` are `foreign keys` and been refereneced from the `species` and `vets` tables. It not necessary to make mention here again that they are primary keys. Kindly take off `PRIMARY KEY (species_id, vet_id));`

File: schema.sql [hunk 3]
```diff
@@ -4,4 +4,23 @@ USE vet_clinic;
...
+-- Create a "join table" called visits to handle this relationship, it should also keep track of the date of the visit. animals and vets
+CREATE TABLE visits (animal_id INTEGER REFERENCES animals(id), vet_id INTEGER REFERENCES vets(id), date DATE, PRIMARY KEY (animal_id, vet_id, date));
```
Comment (t-yanick) [inline schema.sql]: - [ ] Samething like in the `specializations` table above, I don't think `PRIMARY KEY (animal_id, vet_id, date)); `is still very necessary.

File: data.sql [hunk 4]
```diff
@@ -4,4 +4,23 @@ INSERT INTO animals (name, date_of_birth, escape_attempts,neutered,weight_kg) VA
...
...s WHERE name='Plantmon'),(SELECT id FROM vets WHERE name='William Tatcher'),'2020-08-10'), ((SELECT id FROM animals WHERE name='Plantmon'),(SELECT id FROM vets WHERE name='Maisy Smith'),'2021-04-07'), ((SELECT id FROM animals WHERE name='Squirtle'),(SELECT id FROM vets WHERE name='Stephanie Mendez'),'2019-09-29'), ((SELECT id FROM animals WHERE name='Angemon'),(SELECT id FROM vets WHERE name='Jack Harkness'),'2020-08-03'), ((SELECT id FROM animals WHERE name='Angemon'),(SELECT id FROM vets WHERE name='Jack Harkness'),'2020-11-04'), ((SELECT id FROM animals WHERE name='Boarmon'),(SELECT id FROM vets WHERE name='Maisy Smith'),'2019-01-24'), ((SELECT id FROM animals WHERE name='Boarmon'),(SELECT id FROM vets WHERE name='Maisy Smith'),'2019-05-15'), ((SELECT id FROM animals WHERE name='Boarmon'),(SELECT id FROM vets WHERE name='Maisy Smith'),'2020-02-27'), ((SELECT id FROM animals WHERE name='Boarmon'),(SELECT id FROM vets WHERE name='Maisy Smith'),'2021-08-03'), ((SELECT id FROM animals WHERE name='Blossom'),(SELECT id FROM vets WHERE name='Stephanie Mendez'),'2020-05-24'), ((SELECT id FROM animals WHERE name='Blossom'),(SELECT id FROM vets WHERE name='William Tatcher'),'2021-01-11');
```
Comment (t-yanick) [inline data.sql]: **[ OPTIONAL]** 
- You can consider making each **SELECT** statement on a new line to make your codebase readable.

File: queries.sql [hunk 5]
```diff
@@ -80,4 +80,44 @@ SELECT neutered, AVG(escape_attempts) as average_escape FROM animals GROUP BY ne
...
+SELECT species, AVG(escape_attempts) as average_escape FROM animals WHERE date_of_birth BETWEEN '1990-01-01' AND '2000-12-31' GROUP BY species;
+
+/* Vet clinic database: add "join table" for visits */
+-- Write queries to answer the following
+-- Who was the last animal seen by William Tatcher?
+SELECT a.* FROM animals a JOIN visits v ON a.id = v.animal_id JOIN vets vt ON vt.id = v.vet_id WHERE vt.name = 'William Tatcher' ORDER BY v.date DESC LIMIT 1;
+-- How many different animals did Stephanie Mendez see?
+SELECT COUNT(DISTINCT a.id) FROM animals a JOIN visits v ON a.id = v.animal_id JOIN vets vt ON vt.id = v.vet_id WHERE vt.name = 'Stephanie Mendez';
+-- List all vets and their specialties, including vets with no specialties.
+vet_clinic=# SELECT v.name, sp.name FROM vets v  LEFT JOIN specializations s ON v.id = s.vet_id LEFT JOIN species sp ON sp.id = s.species_id ORDER BY v.name; \gdesc
```
Comment (t-yanick) [inline queries.sql]: - [ ] I think this query is commented and cannot run. Kindly check and fix it

File: queries.sql [hunk 6]
```diff
@@ -80,4 +80,44 @@ SELECT neutered, AVG(escape_attempts) as average_escape FROM animals GROUP BY ne
...
+-- What specialty should Maisy Smith consider getting? Look for the species she gets the most.
+SELECT species.name AS specialse_on FROM visits JOIN vets ON visits.vet_id=vets.id JOIN animals ON visits.animal_id=animals.id JOIN species ON animals.species_id=species.id WHERE visits.vet_id = (SELECT id FROM vets WHERE name = 'Maisy Smith') GROUP BY species.name ORDER BY COUNT(visits.animal_id) DESC LIMIT 1;
+
+
+
+
+
+
+
+
```
Comment (t-yanick) [inline queries.sql]: - [ ] Remove these empty lines from your codebase.

**Review by t-yanick — CHANGES_REQUESTED** [review]
Hi @SSEKPIUS , 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good commit messages
:heavy_check_mark: Descriptive Readme file

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by omarsalem7 — APPROVED** [review]
Hi @SSEKPIUS 😊, 
# STATUS: APPROVED! 🥇 💯 

giphy

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 🚀 
Congratulations! 🎉 

## Highlights
✅ Descriptive pull request 
✅ Project has met all the technical aspects you needed for this project. Keep rocking! 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**